---
title: "Mutation statistics"
author: "Markus"
date: "2024-10"
---


```{r}
require(tidyverse)
require(ggplot2)
require(reshape2)
require(forcats)
setwd("Y:/Research/Germline_project/mutations analysis")

# Enrichment across chromosomes
# Repeat masked nucleotide numbers for each chromosome
masked_genome_size <- 34357184
dat <- data.frame(chrom = c("Chr1",
                            "Chr2",
                            "Chr3",
                            "Chr4",
                            "Chr5",
                            "Chr6",
                            "Chr7",
                            "Chr8",
                            "Chr9",
                            "Chr10",
                            "Chr11"),
                  nt_no_rep = c(4975478,
                                4057547,
                                3878176,
                                3237089,
                                3374225,
                                3187697,
                                2896660,
                                2580490,
                                2234235,
                                2063866,
                                1871721),
                  exp_mut = c(33.5366408,27.34943181,26.14040208,21.81922843,
                              22.74357797,21.4863073,19.52460566,17.39349792,
                              15.05960566,13.91125289,12.61612148),
                  obs_mut = c(34,23,27,22,18,25,24,15,18,10,14))
dat$genome_frac <- c(dat$nt_no_rep / masked_genome_size)
chisq <- chisq.test(dat$obs_mut, p= dat$genome_frac)
round(chisq$expected,2)
chisq$observed

# Germline vs soma
dat$mutg <- c(9,5,7,5,3,8,5,4,3,3,3)
dat$muts <- c(25,18,20,17,15,17,19,11,15,7,11)
chisq.test(dat$mutg, dat$muts, simulate.p.value = TRUE)
# p-value = 0.3798

# Plot the mutations per chromosome.
dat_m <- melt(dat[,c(1,7,6)])
colnames(dat_m) <- c("chr","tissue","mutations")
dat_m$Tissue[dat_m$tissue == "mutg"] <- "Germline"
dat_m$Tissue[dat_m$tissue == "muts"] <- "Soma"
dat_m <- dat_m[,c(1,4,3)]
dat_m$Tissue <- factor(dat_m$Tissue, levels = c("Soma", "Germline"))
chr_order <- mixedsort(dat_m$chr)
p <- ggplot(dat_m, aes(x=chr,y=mutations,fill=Tissue)) +
  geom_bar(stat="identity", position="dodge", color = "black") +
  scale_x_discrete(limits=chr_order) +
  ylab("# Mutations") +
  theme_minimal() +
  facet_wrap(~Tissue, nrow=2) +
  theme(axis.title.x=element_blank(),
        legend.position = "none") +
  scale_fill_manual(values=c("#96c176ff", "#cf9ddbff"))
setwd("Y:/Research/Germline_project/mutations analysis/plots")
ggsave(file="mut-per-chr.svg", plot=p, width=4.2, height=3)

# Also normalized by callable chromosome length
dat_norm <- dat_m[,1:2]
dat_norm$norm_mutations <- dat_m$mutations / c(dat$nt_no_rep,dat$nt_no_rep) *1000000
chr_order <- rev(chr_order)
p <- ggplot(dat_norm, aes(x=chr,y=norm_mutations,fill=Tissue)) +
  geom_bar(stat="identity", position="dodge", color = "black") +
  scale_x_discrete(limits=chr_order) +
  ylab("Mutations / Mb") +
  theme_minimal() +
  #facet_wrap(~Tissue, nrow=2) +
  theme(axis.title.x=element_blank(),
        legend.position = "none",
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  scale_fill_manual(values=c("#96c176ff", "#cf9ddbff")) 
  #coord_flip() 
ggsave(file="mut-per-mb.svg", plot=p, width=2.5, height=1.4)

# Plot mutations per chromosome, normalized by length,
# and plot the distance to chromosome to chromosome ends.
read_tsv("mutations.v3.metadata.tsv") |> select(CHROM, pos, ref, alt, ring, `tissue`) -> df_1
# NB: Not the same as the masked length. Got these numbers from the fai
chrom_lengths <- data.frame(unique(df_1$CHROM),
                            c(5997747,5019965,4847850,4278807,4073462,3932117,3741210,3273871,3005252,2796206,2603984)) 
colnames(chrom_lengths) <- c("CHROM", "Length")
# Next, calculate the distance to a chromosome end for each mutation.
calc_dist <- function(crd, chromlen){
  midpoint <- chromlen/2
  absdist <- crd-midpoint
  # Then normalize by chromosome length
  reldist <- absdist/chromlen
  return(reldist)
}
dists_germline <- c()
chr_germline <- c()
for (chrom in chrom_lengths$CHROM){
  chromlen <- chrom_lengths[chrom_lengths$CHROM == chrom,2]
  dat <- df_1[df_1$CHROM == chrom,]
  chr_germline <- c(chr_germline, dat[dat$tissue =="germline",1]$CHROM)
  dat <- dat[dat$tissue =="germline",2]
  dists1 <- lapply(dat, calc_dist, chromlen=chromlen)$pos
  dists_germline <- c(dists_germline,dists1)
}
dists_soma <- c()
chr_soma <- c()
for (chrom in chrom_lengths$CHROM){
  chromlen <- chrom_lengths[chrom_lengths$CHROM == chrom,2]
  dat <- df_1[df_1$CHROM == chrom,]
  chr_soma <- c(chr_soma, dat[dat$tissue =="soma",1]$CHROM)
  dat <- dat[dat$tissue =="soma",2]
  dists1 <- lapply(dat, calc_dist, chromlen=chromlen)$pos
  dists_soma <- c(dists_soma,dists1)
}
df <- data.frame(chrom = c(chr_soma,chr_germline),
                 dist = c(dists_soma,dists_germline),
                 tissue = c(rep("soma",length(chr_soma)),
                            rep("germline",length(chr_germline))))

chromosomes <- c("Chr1","Chr2","Chr3","Chr4","Chr5","Chr6",
                 "Chr7","Chr8","Chr9","Chr10","Chr11")

# Plot cumulative relative distance
# from chromosome midpoint. Since we don't know
# where the centromeres are.
p <- ggplot(df, aes(x = dist, col = tissue)) +
  stat_ecdf(geom = "smooth") +
  geom_abline(intercept=0.5, col = "gray") +
  theme_minimal() +
  scale_color_manual(values=c("#cf9ddbff", "#96c176ff")) +
  theme(legend.position = "none") +
  xlab("Relative distance to chromosome midpoint") +
  ylab("Cumulative fraction of variants")
ggsave(file="cum-dist.svg", plot=p, width=4, height=3)

## CDS vs NCDS
# Numbers obtained by overlapping CDS gff file with mutations vcf file.
dat <- data.frame(obs_mut = c(104, 126),
                  n_bp = c(17324168,17355920))
rownames(dat) <- c("CDS", "NCDS")
chisq.test(dat)
# p-value = 0.1704
tot_bp = 17324168+17355920
pcn_cds = 17324168/tot_bp
pcn_ncds = 17355920/tot_bp
chisq.test(dat$obs_mut, p= c(pcn_cds,pcn_ncds))
# p-value = 0.1508
fisher.test(dat)
# p-value = 0.166

# Germline and soma separately
dat <- data.frame(obs_mut_soma = c(73,102),
                  obs_mut_germline = c(31, 24),
                  n_bp = c(17324168,17355920))
rownames(dat) <- c("CDS", "NCDS")
fisher.test(dat[,c(1,3)]) # 0.03401
fisher.test(dat[,c(2,3)]) # p=0.3492
fisher.test(dat[,c(1,2)]) # p=0.06355

# Let's calculate a "pC/pN"
# for fraction of coding versus fraction of noncoding
# Germline
pCg <- dat$obs_mut_germline[1]/dat$n_bp[1]
pNg <- dat$obs_mut_germline[2]/dat$n_bp[2]
pCg/pNg # 1.294034, so more mutations in CDS than expected.
# Soma
pCs <- dat$obs_mut_soma[1]/dat$n_bp[1]
pNs <- dat$obs_mut_soma[2]/dat$n_bp[2]
pCs/pNs # 0.716998

# Plot the ratios
# Add a hline for the genomic background
bg <- dat$n_bp[2] / (dat$n_bp[1]+dat$n_bp[2])
dat <- dat[,1:2]
dat$region <- rownames(dat)
colnames(dat) <- c("soma", "germline", "region")
dat_m <- melt(dat)
dat_m$region[dat_m$region == "CDS"] <- "coding"
dat_m$region[dat_m$region == "NCDS"] <- "noncoding"
colnames(dat_m) <- c("region","tissue", "mutations")

p <- ggplot(dat_m, aes(x = tissue, y=mutations, fill = region)) +
  geom_bar(position = "fill", stat="identity") +
  theme_minimal() +
  geom_hline(yintercept = bg, linetype = 2, col = "red") +
  scale_fill_manual(values=c("#32527b", "#377650")) +
  theme(axis.title.x=element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  #xlab("Relative distance to chromosome midpoint") +
  ylab("Fraction of mutations")
setwd("Y:/Research/Germline_project/mutations analysis/plots")
ggsave(file="cds-ncds.svg", plot=p, width=2.5, height=1.8)


## Distribution between nucleotypes
# Data from mutations.metadata.v3
ntypes <- rbind(c(62,43),c(18,13))
colnames(ntypes) <- c("A","B")
rownames(ntypes) <- c("soma","germline")
fisher.test(ntypes) # p=1
ntypes_tot <- c(sum(ntypes[1:2,1]),sum(ntypes[1:2,2]))
binom.test(ntypes_tot,p=0.5,alternative="two.sided") # p-value = 0.04818
# Try also the tissues separately.
binom.test(ntypes[1,],p=0.5,alternative="two.sided") # p-value = 0.07849
binom.test(ntypes[2,],p=0.5,alternative="two.sided") # p-value = 0.4731
# Plot it
ntypes <- rbind(c(62,43,70),c(18,13,24))
colnames(ntypes) <- c("Nucleotype A","Nucleotype B", "unphased")
rownames(ntypes) <- c("soma","germline")
dat_m <- melt(ntypes)
colnames(dat_m) <- c("tissue","nucleotype", "mutations")
p <- ggplot(dat_m, aes(x = tissue, y=mutations, fill = nucleotype)) +
  geom_bar(position = "fill", stat="identity") +
  theme_minimal() +
  scale_fill_manual(values=c("#c2261f", "#ffb70b", "gray")) +
  theme(axis.title.x=element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  ylab("Fraction of mutations")
ggsave(file="nucleotypes.svg", plot=p, width=2.5, height=1.8)


## Mutation spectrum
# Boels data & based off her code with some editing.
read_tsv("mutations.v3.metadata.tsv") |> select(CHROM, pos, ref, alt, ring, `tissue`, `CDS/NCDS`) -> df_1

df_1 |> filter( ref == "C" & alt == "A" | ref == "G" & alt == "T") |> mutate( type = "C>A") |> mutate( category = "Tv") -> df_2
df_1 |> filter(ref == "C" & alt == "G" | ref == "G" & alt == "C") |> mutate( type = "C>G") |> mutate( category = "Tv") -> df_3
df_1 |> filter(ref == "C" & alt == "T" | ref == "G" & alt == "A") |> mutate( type = "C>T") |> mutate( category = "Ts") -> df_4
df_1 |> filter(ref == "T" & alt == "A" | ref == "A" & alt == "T") |> mutate( type = "T>A") |> mutate( category = "Tv") -> df_5
df_1 |> filter(ref == "T" & alt == "C" | ref == "A" & alt == "G") |> mutate( type = "T>C") |> mutate( category = "Ts") -> df_6
df_1 |> filter(ref == "T" & alt == "G" | ref == "A" & alt == "C") |> mutate( type = "T>G") |> mutate( category = "Tv") -> df_7
df_1 |> filter(nchar(ref) > nchar(alt)) |> mutate( type = "del") |> mutate( category = "INDEL") -> df_8
df_1 |> filter( nchar(ref) < nchar(alt)) |> mutate(type = "ins") |> mutate(category = "INDEL") -> df_9

###Merge the datasets into one table, and reorder to ensure that indels are put at the end (will make the mutation spectrum figure look nicer)
df_all <- full_join(df_2, df_3) |> full_join(df_4) |> full_join(df_5) |> full_join(df_6) |> full_join(df_7) |> full_join(df_8) |> full_join(df_9) |> arrange(CHROM, pos)

# Ensure the dataframe is ordered correctly
df_all <- df_all |>
  mutate(type = fct_relevel(fct_infreq(type), "del", "ins", after = Inf))

##Step 1: make the mutation spectrum plot
df_all |>
  ggplot(aes(x = type)) +
  geom_bar(position = position_stack(), aes(fill = tissue)) +
  labs(y = "count") +
  scale_fill_manual(values=c("#cf9ddbff", "#96c176ff")) +
    theme_minimal() #just to remove the grey background

# Separate bars and show relative numbers.
# 55 germline mutations, 175 somatic
dat_frac <- data.frame(summary(df_all[df_all$tissue == "soma",]$type)/175)
dat_frac$germline <- summary(df_all[df_all$tissue == "germline",]$type)/55
colnames(dat_frac) <- c("soma", "germline")
dat_frac$type <- rownames(dat_frac)
require(reshape2)
dat_frac <- data.frame(melt(dat_frac))
colnames(dat_frac) <- c("type", "tissue", "frac")
typeorder <- c("C>T","T>C", "C>G","C>A","T>G","T>A","del","ins")
tstv <- data.frame(ts=c(sum(summary(df_all[df_all$tissue == "germline",]$type)[1:2]),
                        sum(summary(df_all[df_all$tissue == "soma",]$type)[1:2])),
                   tv=c(sum(summary(df_all[df_all$tissue == "germline",]$type)[3:6]),
                        sum(summary(df_all[df_all$tissue == "soma",]$type)[3:6])))
tstv_germ <- round(tstv[1,1]/tstv[1,2],1)
tstv_soma <- round(tstv[2,1]/tstv[2,2],1)
p <- ggplot(dat_frac, aes(x = type, y = frac, fill = tissue)) +
  geom_bar(position="dodge", stat="identity", col = "black") +
  scale_fill_manual(values=c("#96c176ff", "#cf9ddbff")) +
  theme_minimal() +
  labs(y ="Fraction of mutations") +
  scale_x_discrete(limits = typeorder) +
  theme(axis.title.x=element_blank(),
        axis.text.x = element_text(face = "bold",family = "Courier New",size = 10,angle = 90, hjust = 1, vjust = 0.5),
        panel.grid.major.x = element_blank(), # Remove major vertical grid lines
        panel.grid.minor.x = element_blank(),  # Remove minor vertical grid lines
        panel.grid.minor.y = element_blank(),  # Remove minor vertical grid lines
        panel.border = element_rect(color = "black", fill = NA, linewidth = 0.5),  # Border around the entire plot
        legend.position = "none") +
  annotate("text",hjust = 1, x=8, y=0.55, label= paste("Germline Ts / Tv",tstv_germ)) +
  annotate("text",hjust = 1, x=8, y=0.50, label= paste("Soma Ts / Tv", tstv_soma))
ggsave(file="mut-spec.svg", plot=p, width=4, height=3)

# Is there a difference between ts and tvs between tissues?
fisher.test(tstv)
# 0.1098

# Test if there is any difference in the spectrum between CDS/NCDS
tmp <- df_all[df_all$tissue == "soma",7:8]
spec_soma <- data.frame(summary(tmp[tmp$"CDS/NCDS" == "CDS",]$type))
spec_soma$NCDS <- summary(tmp[tmp$"CDS/NCDS" == "NCDS",]$type)
colnames(spec_soma) <- c("CDS","NCDS")
tmp <- df_all[df_all$tissue == "germline",7:8]
spec_germline <- data.frame(summary(tmp[tmp$"CDS/NCDS" == "CDS",]$type))
spec_germline$NCDS <- summary(tmp[tmp$"CDS/NCDS" == "NCDS",]$type)
colnames(spec_germline) <- c("CDS","NCDS")

fisher.test(spec_germline) # p-value = 0.293
fisher.test(spec_soma) # p-value = 0.2957
# So there is no difference between coding and non-coding in either tissue.
# Plot it.
spec_germline$tissue <- "germline"
spec_soma$tissue <- "soma"
spec_germline$type <- rownames(spec_germline)
spec_soma$type <- rownames(spec_soma)
spec_both <- rbind(spec_germline,spec_soma)
spec_both_m <- melt(spec_both)
ggplot(spec_both_m, aes(x = type, y = value, fill = tissue)) +
  geom_bar(position="dodge", stat="identity", col = "black") +
  scale_fill_manual(values=c("#cf9ddbff", "#96c176ff")) +
  facet_wrap(variable~tissue) +
  #theme_minimal() +
  labs(y ="Number") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
        

### Selection analysis
# test if the low dnds ratio is significant.
# This analysis was based on only the variant codons.
dnds <- data.frame(germ_obs=c(14,17),
                   soma_obs=c(17,49),
                   germ_sites=c(43/2,143/2),
                   soma_sites=c(142/3,452/3))
rownames(dnds) <- (c("syn","nonsyn"))
# Divide the expected values by three so that the values are comparable.
# Germline
ctable <- cbind(dnds$germ_obs, dnds$germ_sites/3)
fisher.test(ctable) # p=0.1063
# Soma
ctable <- cbind(dnds$soma_obs, dnds$soma_sites/3)
fisher.test(ctable) # p=1
# Tissues against each other
ctable <- cbind(dnds$germ_obs, dnds$soma_obs)
fisher.test(ctable) # 0.06544


## Stats of the SNPeff results.
df <- data.frame(soma = c(8,52,27),
                 germline = c(1,17,16))
rownames(df) <- c("high","moderate", "low")
fisher.test(df) # p=0.2265
chisq.test(df, simulate.p.value = TRUE) # p-value = 0.2264
# Only low-moderate?
df_lomo <- df[2:3,]
chisq.test(df_lomo) # p-value = 0.2277
```